contracts/bridge/BonusChef.sol

```diff
@@ -0,0 +1,640 @@
+// with the following changes:
+// - Expand from single reward token to a list of reward tokens
+// - Allow removing inactive reward pools from list in case list grows above iteration gas limit
+// - Allow governance to rescue unclaimed tokens of inactive pools
+
+// Modified from AladdinDAO MultiStakingRewards contract at https://github.com/AladdinDAO/aladdin-contracts/blob/main/contracts/reward/MultiStakingRewards.sol
+// with the following changes:
+// 1. To ensure compatibility with existing IRewarder interface:
+//      a. SafeERC20 -> BoringERC20
+//      b. onParentReward() -> onSynapseReward()
+//      c. added pendingTokens(), which returns a list of ALL pending rewards for user
+// 2. To ensure compatibility with deployed MiniChefV2 contract:
+//      a. Sending rewards to custom address is possible to make sure
+//         bonus rewards are always transferred to the same address as SYN rewards
+// 3. Removed stake and withdraw, as they happen in the MiniChef
+// 4. Read staking balance and total supply from MiniChef
+// 5. notifyRewardAmount updated to transferFrom(msg.sender) instead of using permissioned distributor
+// 6. Added a few sanity checks
+contract BonusChef is IRewarder, ReentrancyGuard {
+    using SafeMath for uint256;
+    using BoringERC20 for IERC20;
+
+    /* ========== STRUCTS ========== */
+
+    // Info of each reward pool.
+    struct RewardPool {
+        IERC20 rewardToken; // Address of reward token.
+        uint256 periodFinish; // timestamp of when this reward pool finishes distribution
+        uint256 rewardRate; // amount of rewards distributed per unit of time
+        uint256 rewardsDuration; // duration of distribution
+        uint256 lastUpdateTime; // timestamp of when reward info was last updated
+        uint256 rewardPerTokenStored; // current rewards per token based on total rewards and total staked
+        mapping(address => uint256) userRewardPerTokenPaid; // amount of rewards per token already paid out to user
+        mapping(address => uint256) rewards; // amount of rewards user has earned
+        bool isActive; // mark if the pool is active
+    }
+
+    /* ========== STATE VARIABLES ========== */
+
+    address public rewardsDistribution;
+    address public governance;
+
+    IMiniChefV2 public immutable miniChef;
+    uint256 public immutable chefPoolID;
+    IERC20 public immutable chefStakingToken;
+
+    mapping(address => RewardPool) public rewardPools; // reward token to reward pool mapping
+    address[] public activeRewardPools; // list of reward tokens that are distributing rewards
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        IMiniChefV2 _miniChef,
+        uint256 _chefPoolID,
+        address _rewardsDistribution
+    ) public {
+        miniChef = _miniChef;
+        chefPoolID = _chefPoolID;
+        chefStakingToken = _miniChef.lpToken(_chefPoolID);
+        rewardsDistribution = _rewardsDistribution;
+        governance = msg.sender;
```

Comment:
Even though each "role" seems to only have one address, using something like `AccessControl` makes more sense from a "security-tested" standpoint.

Reply:
Was going to make the same comment. Think this is the standard minichef implementation/the goal here is to have one role:one address. @ChiTimesChi?

Reply:
The initial goal was to modify the already working 3rd party contracts with as little modifications as possible. Also, as there's one address per role and not too many roles, I haven't bothered reworking this into `AccessControl`. 

Nonetheless, it might make sense to do so to avoid future situations where we need more than one `rewardsDistribution` address. Don't have a strong opinion here, I'll trust your gut @trajan0x @bleaker0x

Reply:
I'm generally a fan of `AccessControl` and (IMO) it's fairly painless to implement, but that's my opinion. From a future proofing perspective (as you pointed out), as well as the general ease of implementation, usage, and relative lack of gas inefficiency (yes I know having however many layers layers of internal function calls isn't _great_ but it's a hell of a lot better than some other abstract contracts), I'm pro having it here. Ultimately, your call.